### PR TITLE
Use generator for walking

### PR DESCRIPTION
**Pull request by jrandolf**
Generators generate a better API for traversing trees. This PR also cleans up the callback logic in `walk` code.

### CODE REVIEW

File: parsel.ts [hunk 1]
```diff
@@ -406,33 +429,42 @@ export function specificity(selector: string | AST): number[] {
...
+			case 'id':
+				ret[0]++;
+				break;
+			case 'class':
+			case 'attribute':
+				ret[1]++;
+				break;
+			case 'pseudo-element':
+			case 'type':
+				ret[2]++;
```
Comment (LeaVerou) [inline parsel.ts:444]: This is not an equivalent condition: note that the original code also checks `node.content` if `node.type` is `"type"`.

**Review by LeaVerou — CHANGES_REQUESTED** [review]
I *really* like the idea of using generators, I've been looking for a chance to try them out!
Not sure about rewriting the long conditional in L411 as a switch, since it's not only keying off a specific property, but checks multiple things, so it becomes awkward to rewrite as a switch as you have to separate the conditions into those on `token.type` and rewrite the rest as ifs (not to mention the inherent awkwardness of `break` all over the place). Note that the new code is not completely equivalent to the old, as the conditional in L420 could not be rewritten elegantly with a switch. I'd suggest we stick to the `if` unless you feel very strongly about it.

I'm also not sure why to change `node` to `token`?

File: parsel.ts (hunk 1 above)
Comment (jrandolf) [reply parsel.ts:444]: Because of the `universal` types are distinct type now, we can get rid of that condition :)

Comment (jrandolf) [reply parsel.ts:444]: > I'm also not sure why to change node to token?

By habit. For this case, I named it token so it's clear we are handling tokens, not nodes of an AST.

**Comment by LeaVerou** [comment]
I’m a bit confused about what happened here, it appears this was merged before review, then there's a review requesting changes? How is this possible, I don't think this was merged when I reviewed?

**Pull request merged**